pkg/agent/reset.go

```diff
@@ -0,0 +1,223 @@
+package agent
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/osquery/runsimple"
+	"go.etcd.io/bbolt"
+)
+
+var (
+	hostDataKeySerial       = []byte("serial")
+	hostDataKeyHardwareUuid = []byte("hardware_uuid")
+	hostDataKeySecret       = []byte("secret")
+)
+
+// ResetDatabaseIfNeeded checks to see if the hardware this installation is running on
+// has changed, by checking current hardware-identifying information against stored data
+// in the HostDataStore. If the hardware-identifying information has changed, it performs
+// a backup of the database, and then clears all data from it.
+func ResetDatabaseIfNeeded(ctx context.Context, k types.Knapsack) {
+	k.Slogger().Info("checking to see if database should be reset...")
+
+	serialChanged := false
+	hardwareUUIDChanged := false
+	currentSerial, currentHardwareUUID, err := currentSerialAndHardwareUUID(ctx, k)
+	if err != nil {
+		k.Slogger().Warn("could not get current serial and hardware UUID", "err", err)
+	} else {
+		serialChanged = valueChanged(k, currentSerial, hostDataKeySerial)
+		hardwareUUIDChanged = valueChanged(k, currentHardwareUUID, hostDataKeyHardwareUuid)
+	}
+
+	enrollSecretChanged := false
+	currentEnrollSecret, err := currentSecret(k)
+	if err != nil {
+		k.Slogger().Warn("could not get current enroll secret", "err", err)
+	} else {
+		enrollSecretChanged = valueChanged(k, currentEnrollSecret, hostDataKeySecret)
+	}
+
+	if serialChanged || hardwareUUIDChanged || enrollSecretChanged {
+		k.Slogger().Warn("detected new hardware or rollout, backing up and resetting database",
+			"serial_changed", serialChanged,
+			"hardware_uuid_changed", hardwareUUIDChanged,
+			"enroll_secret_changed", enrollSecretChanged)
+
+		if err := takeDatabaseBackup(k); err != nil {
+			k.Slogger().Warn("could not take database backup", "err", err)
+		}
+
+		wipeDatabase(k)
+
+		// Cache data for future checks
+		if err := k.HostDataStore().Set(hostDataKeySerial, []byte(currentSerial)); err != nil {
+			k.Slogger().Warn("could not set serial in host data store", "err", err)
+		}
+		if err := k.HostDataStore().Set(hostDataKeyHardwareUuid, []byte(currentHardwareUUID)); err != nil {
+			k.Slogger().Warn("could not set hardware UUID in host data store", "err", err)
+		}
+		if err := k.HostDataStore().Set(hostDataKeySecret, []byte(currentEnrollSecret)); err != nil {
+			k.Slogger().Warn("could not set secret in host data store", "err", err)
+		}
+	}
+}
+
+// currentSerialAndHardwareUUID queries osquery for the required information.
+func currentSerialAndHardwareUUID(ctx context.Context, k types.Knapsack) (string, string, error) {
```

Comment:
I know we can get the serial using `dmidecode` on Linux and `wmic bios get serialnumber` on Windows because that's what I do in some of the automated tests, and it looks like `ioreg` will do that for us on MacOS, so we could exec those here, if we wanted. If there's a built-in for Golang I'm not aware of it.

Since we retrieve the serial and hardware UUID from osquery when we're getting enrollment details, I liked doing it the same way here for consistency. But since this blocks rungroup startup and therefore launcher startup, I could see wanting to pick the fastest possible option, too.